### ISSUE TITLE
Evaluate inference support for arduino as arduino lib

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#267]
# Findings from SKaiNET Repo (develop branch)

## Project Structure

The SKaiNET repository is organized as a Kotlin multiplatform project with multiple modules. Key modules include:

### `skainet-lang`
- Contains the core DSL and model definitions (published as **`skainet-lang-core`**).
- Users define networks using a Kotlin DSL. Example: a simple MLP:

```kotlin
val model = nn {
    input(64)
    dense(out = 64)
    dense(out = 10)
}
````

* In this DSL:

  * `nn { ... }` likely constructs a neural network object
  * `input(...)` specifies input tensor dimensions
  * `dense(out = N)` adds a Dense (fully-connected) layer with **N outputs**
* The DSL is declarative: it builds a **model description** rather than executing immediately.
* Internally, SKaiNET defines layer classes (e.g. `Dense`, `Conv2D`, `Dropout`) and a builder to accumulate them.
* The `skainet-lang-core` module provides these layer classes and a `NeuralNetwork` container class holding layers and weights.

### `skainet-backends`

* Provides execution backends (e.g. **`skainet-backend-cpu`**).
* The default CPU backend implements operations for training/inference on JVM/Native.
* Likely includes matrix multiplication, activation functions, etc., optimized per platform (README mentions SIMD usage on JVM).
* The CPU backend consumes the DSL-defined model and runs forward/backward passes.
* Uses a tensor / ND-array structure for data.

### `skainet-compile`

* Contains compiler or IR logic.
* Appears to house Kotlin Symbol Processing (KSP) components:

  * Maven artifacts include `skainet-lang-ksp-annotations` and `skainet-lang-ksp-processor` (per Klibs.io).
* Suggests SKaiNET uses annotations + KSP processor to analyze/transform DSL definitions at compile-time.
* Possible uses: boilerplate generation, static verification (e.g., shape checking).
* Architecture diagram (0.5.0) hints at a compiler pipeline.
* As of develop branch: **no existing TinyML C code generation**; KSP is likely internal graph building/validations rather than emitting C.

### `skainet-io`

* Handles model serialization/loading.
* Supports reading models in **GGUF** format and **JSON**.
* Implies an internal representation of model weights that can be saved/loaded across platforms.

### `skainet-data`

* Provides data loading and dataset APIs (e.g. an MNIST loader).
* Not directly about codegen, but relevant for training models in Kotlin.

### `skainet-models`

* Contains pre-built model definitions (e.g. `skainet-model-yolo`).
* Likely uses the DSL to construct known architectures (YOLO, etc.).
* Serves as a “model zoo” (`skainet-lang-models` includes simple examples).

---

## DSL Entry Points

* Primary entry point: `nn { ... }` builder (likely a top-level DSL function).
* Inside the DSL: layer construction functions like `input(...)`, `dense(...)`, `conv2d(...)`, etc.
* Under the hood: likely extension functions on a `NeuralNetworkBuilder` context.
* Example behavior:

  * `dense(out = 64)` creates a Dense layer object, sets output size to 64, and appends it to the model’s layer list.
* Activations may be:

  * parameters on layers, or
  * separate calls (e.g. `relu()` after a layer), depending on implementation.

---

## Graph/IR Representation

* SKaiNET currently looks eager and object-based:

  * a sequence of layer objects rather than a standalone IR graph.
* Each layer likely holds parameters (weights/biases) and implements `forward()` (and for training, also backward).
* Example usage implies something like `model.forward(x, ctx)` to run forward pass.
* Presence of `DefaultNeuralNetworkExecutionContext` and `train { ... }` / `eval { ... }` suggests training vs inference is driven by context:

  * e.g. Dropout/BatchNorm behave differently depending on context mode.
* No explicit separate IR is mentioned in develop branch; the “IR” is effectively the Kotlin layer objects.
* KSP implies some compile-time intermediate form could exist, but likely for checks/registration rather than emitting C.

---

## Compilation / Codegen Hooks

* Aside from KSP, docs include an architecture diagram suggesting a compiler-inspired lowering approach.
* SKaiNET is multiplatform (JVM, Native, JS). CPU backend likely has variants per platform.
* `kotlin-js-store` suggests special handling for weights in JS contexts.
* No module generates C currently, but `skainet-compile` is a natural place to add TinyML codegen (or to create new modules).

---

## Training vs Inference Structure

* SKaiNET supports training/backprop on at least some backends.
* Layers likely implement forward and backward. Dense backward computes grads for weights + inputs.
* Weights after training live in the layer objects, so export requires:

  * extracting final weights/biases from trained model.

---

## Relevant Classes / Files (inferred)

Without direct code, likely candidates include:

* `skainet-lang-core`: `NeuralNetwork` / `SequentialModel`, `Layer` base type, concrete layers (`DenseLayer`, `Conv2DLayer`, etc.), DSL builders (`nn{}`, `input()`, `dense()`, ...).
* `skainet-lang-ksp-annotations`: annotations such as `@Model`, `@Layer` (speculative).
* `skainet-lang-ksp-processor`: generates code for shape checking, or stubs, etc. (speculative).
* `skainet-backend-cpu`: math ops like matmul, activations; weights stored in arrays or platform tensors.
* `skainet-compile`: KSP coordination and potentially any IR definitions; could be extended with TinyML pipeline.

---

# TinyML Comparison and Lessons

## TensorFlow Lite for Microcontrollers (TFLM)

* Uses an interpreter approach with a `.tflite` FlatBuffer model.
* Model stored as a C array (memory-mapped) and executed op-by-op.
* FlatBuffer contains a topologically sorted op list (avoids runtime graph traversal).
* Runtime still translates FlatBuffer parameters/tensors into ops during inference → overhead exists.
* **Memory planning:** static *tensor arena* allocated by user; interpreter pre-allocates tensors/scratch during init.

  * Two-stack strategy: long-lived allocations from one end, temporaries from the other.
  * Can reuse buffers for non-overlapping tensor lifetimes.
* Strong quantization support (int8) for footprint + speed.
* Optimized kernels exist for many ops (often via CMSIS-NN on ARM).
* Arduino usage: include model array + interpreter, call `Invoke()`.
* Lesson: interpreter overhead + larger binary; model-specific generated code can avoid overhead.

## STM32Cube.AI (X-Cube-AI)

* Ahead-of-time C code generator for STM32.
* Generates C source/header embedding weights and layer calls.
* Compile-time memory optimization and usage reports; predictable allocation.
* Supports float32 and INT8; can use CMSIS-NN kernels when quantized.
* Often faster/smaller than interpreter-based approaches by tailoring code.
* Provides validation workflows to compare generated outputs to original.
* Lesson: codegen + validation is a strong pattern for TinyML deployment.

## Emlearn (Embedded ML)

* Generates portable, self-contained C99 inference code for embedded devices.
* Emphasizes:

  * no dynamic allocation
  * very small RAM/code footprint
* Supports simple neural nets (especially sequential Dense/MLP).
* Can generate:

  * single-header “inline” model with weights + predict function, or
  * model-struct + generic runtime.
* Distributed as Arduino library too.
* Lesson: simple MLP codegen is effective and aligns with SKaiNET phase-1 scope.

## General Lessons

* Static generated code tends to be smaller/faster than interpreter runtimes.
* Compile-time memory planning is crucial.
* Quantization is valuable but adds complexity; float32 first is reasonable.
* Start narrow: Dense + common activations.
* Arduino packaging matters (library structure + example sketch).

---

# Proposed SKaiNET C Codegen Architecture

## A. Intermediate Representation (IR)

Introduce a minimal IR to represent the network in a backend-agnostic way, used as the input for C codegen.

### IR components

* **Tensor**: fixed shape + dtype (initially float32).
* **Constant**: tensor with constant data (weights/biases).
* **Op**:

  * `Dense/FullyConnected`: input vector → output vector using W and b
  * `Activation`: ReLU / Sigmoid / Tanh / (optional) Softmax
* Graph: DAG, but for MLP it's a simple chain.

### Example IR for 2-layer MLP with ReLU

* Input `X [in_dim]`
* Constants: `W1 [hidden×in]`, `b1 [hidden]`, `W2 [out×hidden]`, `b2 [out]`
* Ops:

  * `Dense(W1,b1): X -> H`
  * `ReLU: H -> H2`
  * `Dense(W2,b2): H2 -> Y`
  * Optional output activation

### Suggested Kotlin data model (sketch)

```kotlin
data class Tensor(val name: String, val shape: IntArray, val dtype: DType)
data class Constant(val tensor: Tensor, val data: FloatArray)

sealed interface Op {
    data class Dense(val weight: Constant, val bias: Constant, val input: Tensor, val output: Tensor) : Op
    data class Activation(val type: ActType, val input: Tensor, val output: Tensor) : Op
}

data class Graph(val inputs: List<Tensor>, val outputs: List<Tensor>, val ops: List<Op>)
```

### Lowering from SKaiNET model to IR

* Implement in `skainet-compile` (or new module e.g. `skainet-compile-c`).
* Function: `MLPIRCompiler.compile(model: NeuralNetwork): Graph`
* Steps:

  * validate supported layers (Dense + supported activations)
  * create input tensor from `input()`
  * walk layer list, create constants for weights/bias
  * create output tensors for each op
  * mark final tensor as graph output
* Requires weight extraction API (add if missing).

---

## B. Code Generation (IR → C)

Generate:

* `model.h`
* `model.c`

### Constants

* Emit weights/biases as `static const float` arrays (flash/ROM).
* Naming: `MODEL_W0`, `MODEL_B0`, `MODEL_W1`, `MODEL_B1`, ...
* Emit dimension macros:

  * `#define MODEL_INPUT_DIM ...`
  * `#define MODEL_OUTPUT_DIM ...`

### Inference API

* Function signature:

  * `void model_infer(const float *input, float *output);`

### Memory strategy

* No malloc.
* Use two buffers sized to maximum intermediate width:

  * ping-pong between `buf_a` and `buf_b`.
* Optionally write final output directly to `output`.

### Dense op emission

```c
for (int j = 0; j < OUT_DIM; ++j) {
    float acc = 0.0f;
    for (int i = 0; i < IN_DIM; ++i) {
        acc += in[i] * W[j*IN_DIM + i];
    }
    acc += B[j];
    out[j] = acc;
}
```

### Activation emission

* ReLU: clamp at 0
* Sigmoid/Tanh: use `<math.h>` functions (`expf`, `tanhf`) for Phase 1
* Softmax: stable implementation using max subtraction + exp

### Headers

* `extern "C"` guards for Arduino/C++ integration
* Expose only inference function and dimension macros

---

### Export workflow

* Programmatic API preferred (weights known after training):

  * `SkainetExporter.exportToC(model, "MyModel", outputDir)`
* Gradle task optional later; KSP-based export is harder due to trained weights.

### Arduino library output structure

```
Skainet_MyModel/
  library.properties
  src/
    MyModel.c
    MyModel.h
  examples/
    MyModelTest/
      MyModelTest.ino
```

### Sketch usage example

```cpp
#include "MyModel.h"

void setup() {
  Serial.begin(115200);
  float input[MODEL_INPUT_DIM] = { /* ... */ };
  float output[MODEL_OUTPUT_DIM];
  model_infer(input, output);
  Serial.println(output[0]);
}

void loop() {}
```

### AVR / PROGMEM note

* Phase 1: keep code simple, no automatic PROGMEM.
* Potential future enhancement: `#ifdef __AVR__` + PROGMEM + `pgm_read_float`.

---

### Golden-output tests

* Generate random + edge-case inputs.
* Compare:

  * SKaiNET Kotlin output (`model.forward`)
  * Host-compiled C output (`model_infer`)
* Use tolerance (e.g., abs/rel error < 1e-6), especially for sigmoid/tanh/softmax.

### CI ideas

* Run codegen on a sample model.
* Compile generated C with gcc on host and run tests.
* Optionally compile with `arduino-cli` for a few boards to catch toolchain issues.

---

### Phase 1 (current scope)

* Float32 MLP export
* Dense + basic activations
* Static allocation, portable C99
* Arduino library packaging + tests

### Phase 2

* INT8 quantization support
* quant params in IR, int8 weights, int32 bias accumulations
* optional calibration / PTQ pipeline

### Phase 3

* Optional optimized kernels (CMSIS-NN) for ARM targets

### Phase 4

* Expand ops (Conv2D, pooling, etc.)

### Phase 5 (optional)

* batch support, fusion/in-place optimizations, better tooling

---

# Open Questions & Recommended Defaults

* **Naming:** index-based (`W0`, `B0`...) by default; add comments mapping indices to layers.
* **Integration:** programmatic `exportToC()` default.
* **Weights retrieval:** expose weights/bias as `FloatArray` via API changes if needed.
* **Memory:** two buffers for clarity in Phase 1; consider in-place later.
* **AVR PROGMEM:** not default in Phase 1; document/option later.
* **Math:** use `<math.h>` functions for correctness; mention speed tradeoffs.
* **Input scaling:** no implicit scaling; user must match training preprocessing.

---

# Sources (as provided)

## Klibs

* sk.ainet.core:skainet-lang-ksp-processor:0.5.0 — Klibs.io
  [https://klibs.io/package/sk.ainet.core/skainet-lang-ksp-processor](https://klibs.io/package/sk.ainet.core/skainet-lang-ksp-processor)

## GitHub

* SKaiNET repository
  [https://github.com/sk-ai-net/skainet/](https://github.com/sk-ai-net/skainet/)
* emlearn repository
  [https://github.com/emlearn/emlearn](https://github.com/emlearn/emlearn)
* MLP-for-ESP32
  [https://github.com/lesept777/MLP-for-ESP32](https://github.com/lesept777/MLP-for-ESP32)

## Papers / Docs

* TensorFlow Lite Micro: Embedded Machine Learning on TinyML Systems
  [https://arxiv.org/pdf/2010.08678](https://arxiv.org/pdf/2010.08678)
* TFLM memory management (Zread)
  [https://zread.ai/tensorflow/tflite-micro/9-memory-management-system](https://zread.ai/tensorflow/tflite-micro/9-memory-management-system)
* STM32Cube.AI
  [https://stm32ai.st.com/stm32-cube-ai/](https://stm32ai.st.com/stm32-cube-ai/)
* ST wiki: X-CUBE-AI CLI automation
  [https://wiki.st.com/stm32mcu/wiki/AI:How_to_automatize_code_generation_and_validation_with_X-CUBE-AI_CLI](https://wiki.st.com/stm32mcu/wiki/AI:How_to_automatize_code_generation_and_validation_with_X-CUBE-AI_CLI)
* emlearn Arduino getting started
  [https://emlearn.readthedocs.io/en/latest/getting_started_arduino.html](https://emlearn.readthedocs.io/en/latest/getting_started_arduino.html)
